modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/MMDBUtil.java

```diff
@@ -0,0 +1,101 @@
+     * loading this data would otherwise be wasteful if they are not also ingest nodes.
+     *
+     * @return the database type
+     * @throws IOException if an I/O exception occurs reading the database type
+     */
+    public static String getDatabaseType(final Path database) throws IOException {
+        final long fileSize = Files.size(database);
+        try (InputStream in = Files.newInputStream(database)) {
+            // read the last BUFFER_SIZE bytes (or the fileSize, whichever is smaller)
+            final long skip = fileSize > BUFFER_SIZE ? fileSize - BUFFER_SIZE : 0;
+            final long skipped = in.skip(skip);
+            if (skipped != skip) {
+                throw new IOException("failed to skip [" + skip + "] bytes while reading [" + database + "]");
+            }
+            final byte[] tail = new byte[BUFFER_SIZE];
+            int read = 0;
+            int actualBytesRead;
+            do {
+                actualBytesRead = in.read(tail, read, BUFFER_SIZE - read);
+                read += actualBytesRead;
+            } while (actualBytesRead > 0);
+
+            // find the database_type header
+            int metadataOffset = -1;
+            int markerOffset = 0;
+            for (int i = 0; i < tail.length; i++) {
+                byte b = tail[i];
+
+                if (b == DATABASE_TYPE_MARKER[markerOffset]) {
+                    markerOffset++;
+                } else {
+                    markerOffset = 0;
+                }
+                if (markerOffset == DATABASE_TYPE_MARKER.length) {
+                    metadataOffset = i + 1;
+                    break;
+                }
+            }
+
+            if (metadataOffset == -1) {
+                throw new IOException("database type marker not found");
+            }
+
+            // read the database type
+            final int offsetByte = fromBytes(tail[metadataOffset]);
+            final int type = offsetByte >>> 5;
+            if (type != 2) { // 2 is the type indicator in the mmdb format for a UTF-8 string
+                throw new IOException("type must be UTF-8 string");
+            }
+            int size = offsetByte & 0x1f;
+            if (size == 29) {
+                // then we need to read in yet another byte and add it onto this size
+                // this can actually occur in practice, a 29+ character type description isn't that hard to imagine
+                size = 29 + fromBytes(tail[metadataOffset + 1]);
+                metadataOffset += 1;
+            } else if (size >= 30) {
+                // we'd need to read two or three more bytes to get the size, but this means the type length is >=285
+                throw new IOException("database type marker not found [size == " + size + "]");
+            }
+
+            return new String(tail, metadataOffset + 1, size, StandardCharsets.UTF_8);
+        }
+    }
+
+    static int fromBytes(byte b1) {
```

Comment:
Looks like this could be private right? If not, we probably ought to put some comments on it.

Reply:
607502b